### PR TITLE
[feature/fromAsync] add from async for single

### PR DESCRIPTION
**Pull request by MortyMerr**
I think the current implementation of `fromAsync` might be a little more useful. Completion handlers used to be `(result: T, error: CustomError)` closures. Also completion handler should emit only once. So i used `Single` to implement that

### CODE REVIEW

**Comment by patmalt** [comment]
+1

File: Tests/RxSwift/fromAsyncTests.swift [hunk 1]
```diff
@@ -36,4 +44,47 @@ class FromAsyncTests: XCTestCase {
 
         XCTAssertEqual(observer.events, correct)
     }
+    
+    func testSingleResultEqualitySuccessCase() {
+        // given
+        let result = "result"
+        let expectedEvents = [next(0, result), completed(0)]
```
Comment (freak4pc) [inline Tests/RxSwift/fromAsyncTests.swift]: Please don't use the global deprecated `next` and `completed` functions

File: Cartfile.resolved [hunk 2]
```diff
@@ -1 +1 @@
-github "ReactiveX/RxSwift" "4.4.0"
+github "ReactiveX/RxSwift" "4.4.1"
```
Comment (freak4pc) [inline Cartfile.resolved]: No reason to bump version here

File: Source/RxSwift/fromAsync.swift [hunk 3]
```diff
@@ -66,3 +66,71 @@ extension Observable {
         return { (a: A, b: B, c: C, d: D, e: E, f: F, g: G, h: H, i: I, j: J) in Observable.fromAsync(curry(asyncRequest)(a)(b)(c)(d)(e)(f)(g)(h)(i)(j)) }
     }
 }
+
+public enum FromAsyncError: Error {
+    /// Both result & error can't be nil
+    case inconsistentCompletionResult
```
Comment (freak4pc) [inline Source/RxSwift/fromAsync.swift]: Is this relevant to Observable fromAsync as well ? Why can't result be optional and nil?

Comment (MortyMerr) [reply Source/RxSwift/fromAsync.swift]: Hi @freak4pc . Thank u for ur comments. Do u have any idea how should I handle a situation in which completionHandler have both error and success? I think throwing an inconsistent state error is the best option.

Comment (MortyMerr) [reply Source/RxSwift/fromAsync.swift]: If user want to make result optional type, fromAsync generic type would be double optional
`A??`, so it will work ok

File: Cartfile.resolved (hunk 2 above)
Comment (MortyMerr) [reply Cartfile.resolved]: Sure, fixed

File: Tests/RxSwift/fromAsyncTests.swift (hunk 1 above)
Comment (MortyMerr) [reply Tests/RxSwift/fromAsyncTests.swift]: I can't see any warning about deprecated methods usage. I also can't find deprecation in docs. Could u please attach some proof, that this functions were deprecated?

File: Source/RxSwift/fromAsync.swift (hunk 3 above)
Comment (MortyMerr) [reply Source/RxSwift/fromAsync.swift]: I mean, outer optional goes for result optionality and inner optional is user defined type

File: Tests/RxSwift/fromAsyncTests.swift (hunk 1 above)
Comment (mkko) [reply Tests/RxSwift/fromAsyncTests.swift]: It's a bit implicit, but here: [RxTest/Deprecated.swift](https://github.com/ReactiveX/RxSwift/blob/master/RxTest/Deprecated.swift#L35)

Comment (freak4pc) [reply Tests/RxSwift/fromAsyncTests.swift]: The deprecations are runtime deprecations but they've been there for a while.

File: Source/RxSwift/fromAsync.swift (hunk 3 above)
Comment (mkko) [reply Source/RxSwift/fromAsync.swift]: I wouldn't treat empty result as an error nevertheless.

Comment (MortyMerr) [reply Source/RxSwift/fromAsync.swift]: Why? This is the way, how single works. It should fire exactly one event OR one error. I thunk throwing this documented error would be correct.
I agree about other comments, but here I am pretty sure in implementation

Comment (MortyMerr) [reply Source/RxSwift/fromAsync.swift]: Do u have any suggestions? Maybe better way to implement that?

Comment (freak4pc) [reply Source/RxSwift/fromAsync.swift]: @mkko I'm not too familiar with this operator, so if you've used it and have some opinions it would be highly helpful.

Comment (mkko) [reply Source/RxSwift/fromAsync.swift]: Sorry for the delay. I think that's a fair point, it does look like SingleAsync also expects a value and conformity should be prioritized.

I'm just thinking about a use case where you're not expecting a value, health/auth checks and such. Does it work if you define a `Single` with an optional value?

Comment (MortyMerr) [reply Source/RxSwift/fromAsync.swift]: I fully understand your concerns. But completion handler which has optional value should look like that:
```
typealias ResulType = String? 
func foo(completionHandler: (ResultType?, TestError?) -> Void)
// which is equal to
func foo(completionHandler: (String??, TestError?) -> Void)
```
We use outer optional to show that completion can be called with result OR with an error. And our inner optional goes for our actual result (which type is `String?`l).
About auth check - user should use `Completable` instead. But if he wants to use `Single` its `ElementType` type should be `Void`

Comment (mkko) [reply Source/RxSwift/fromAsync.swift]: Fair point. I think this looks good 👍

Comment (freak4pc) [reply Source/RxSwift/fromAsync.swift]: That makes sense - could we add a quick unit test for the optional case, just for sanity and coverage? I'll merge after that.

File: Tests/RxSwift/fromAsyncTests.swift (hunk 1 above)
Comment (MortyMerr) [reply Tests/RxSwift/fromAsyncTests.swift]: Thanks, resolved

File: Tests/RxSwift/fromAsyncTests.swift [hunk 4]
```diff
@@ -36,4 +44,47 @@ class FromAsyncTests: XCTestCase {
 
         XCTAssertEqual(observer.events, correct)
     }
+    
+    func testSingleResultEqualitySuccessCase() {
```
Comment (freak4pc) [inline Tests/RxSwift/fromAsyncTests.swift]: basically another test case where you do `let result: String? = "result"` would be great. The rest of the test should probably remain the same

File: Source/RxSwift/fromAsync.swift (hunk 3 above)
Comment (MortyMerr) [reply Source/RxSwift/fromAsync.swift]: Just added additional test case with optional result type

**Comment by MortyMerr** [comment]
@freak4pc is it ok now?

File: Tests/RxSwift/fromAsyncTests.swift [hunk 5]
```diff
@@ -36,4 +44,70 @@ class FromAsyncTests: XCTestCase {
...
+            .asObservable()
+            .subscribe(observer)
+        scheduler.start()
+        // then
+        XCTAssertEqual(observer.events, expectedEvents)
+    }
+    
+    func testSingleOptionalResultEqualitySuccessCase() {
+        // given
+        let result: String? = "result"
```
Comment (mkko) [inline Tests/RxSwift/fromAsyncTests.swift]: I would test this against a `nil` value instead of string, since it would be the exceptional case. Otherwise this looks good to me.

**Comment by mkko** [comment]
Just one thing about the testing of optional value. What do you think @freak4pc?

File: Tests/RxSwift/fromAsyncTests.swift (hunk 5 above)
Comment (MortyMerr) [reply Tests/RxSwift/fromAsyncTests.swift]: Changed

**Comment by MortyMerr** [comment]
Hey @mkko thank you for your review) Waiting for final approve

**Comment by mkko** [comment]
Thank you, especially for the patience!
